pep-9999.rst

```diff
@@ -0,0 +1,257 @@
+PEP: 9999
+Title: Package index configuration standardization
+Author: Fredrik Averpil <fredrik at averpil.com>
+Status: Draft
+Type: Informational
+Created: 05-May-2021
+Post-History: <dates of postings to python-ideas and/or python-dev>
+
+
+Abstract
+========
+
+This PEP standardizes configuration of PyPI repositories.
+
+
+Motivation
+==========
+
+Workflow tools come up with their own configuration mechanisms for PyPI repositories, as there is no standard for this. This causes fragmentation and pain points.
+
+Examples of such configuration files:
+
+* Pip's ``pip.conf`` and ``pip.ini``
+* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
+* Poetry's ``config.toml`` and ``auth.toml``
+
+
+Rationale
+=========
+
+While this PEP aim to come up with a unified PyPI repository configuration, thought has been given so to be able to extend the configuration later. For example, PyPI repositories are assumed in this PEP, but a ``type`` attribute could be introduced which would default to ``pypi`` down the line, to allow for other package repository types. Any tool could also start taking advantage of the fact that the configuration files and their discovery is a standard, and use this for their own configuration.
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for PyPI repository configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file::
+
+
+    [repositories]
+
+    [repostiories.pypi]
+    url = "https://pypi.org/simple/"
+
+    [repostiories.myrepo1]
+    url = "http://myserver2.org/simple"
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+A repository has the following attributes::
+
+    ==================  ===========  =============================================
+    Attribute           Type         Description
+    ==================  ===========  =============================================
+    final               bool         Make it impossible to override/remove the repository configuration (default: false)
+    priority            int          Order of priority, sorted descending (default: 0)
+    proxy_url           str          The URL to the proxy server, can contain a port
+    shadow              bool         Use package from this index, even if it was found in a previous index (default: false)
```

Comment:
Does this offer much value? I am not clear on what scenario this really addresses. Further "shadow" implies "fallback" (to me) and the instructions "Use package from this index" implies this takes priority. Again, this caues fair amount of reasoning issues and can also have concerns around PEP 610 and lock files. Feels like the config is attempting to dictate the behaviour of downstream tools.

Reply:
I agree, `shadow` doesn't really add anything with `priority`.

Reply:
I'm not sure what I think about shadowing. Not something I would use personally. It was raised in the [Discourse thread](https://discuss.python.org/t/adding-a-global-config-to-specify-package-indexes/8599/32).

I will remove it. The question is whether we shall put it under "rejected ideas" or not. I will do that for now, with your reasoning behind it @abn.